### PR TITLE
[MLIR][XLA] Add GatherOp to LHLO/HLO emitters

### PR DESCRIPTION
**Pull request by xinan-jiang**
This is a PR from JIZHI, the AI platform in Tencent.
@sherhut @pifon2a

We work on TensorFlow/MLIR to make mlir_gpu enable. Could you tell me how to run the test? It seems need a new tool.

### CODE REVIEW

**Comment by joker-eph** [comment]
@timshen91 can you help providing guidance about where we want the HLO->LHLO conversion be handled right now?

File: tensorflow/compiler/xla/service/mlir_gpu/tests/gather.hlo [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  %x = f32[100,10] parameter(0)
+  %y = s64[4,6] parameter(1)
+  ROOT %gather = f32[4,6,10]{2,1,0} gather(f32[100,10]{1,0} %x, s64[4,6]{1,0} %y),
+      collapsed_slice_dims={0}, index_vector_dim=2, offset_dims={2},
+      slice_sizes={1,10}, start_index_map={0}
+}
+
+// CHECK: func @gather(%[[ARG0:.*]]: [[TYPE0:.*]], %[[ARG1:.*]]: [[TYPE1:.*]], %[[RESULT:.*]]: [[RTYPE:.*]]) {
+// CHECK:   "xla_lhlo.gather"(%[[ARG0]], %[[ARG1]], %[[RESULT]])
+// CHECK:   {collapsed_slice_dims = dense<0> : tensor<1xi64>, index_vector_dim = 2 : i64, offset_dims = dense<2> : tensor<1xi64>, slice_sizes = dense<[1, 10]> : tensor<2xi64>, start_index_map = dense<0> : tensor<1xi64>} : ([[TYPE0]], [[TYPE1]], [[RTYPE]]) -> ()
```
Comment (pifon2a) [inline tensorflow/compiler/xla/service/mlir_gpu/tests/gather.hlo:14]: nit: please use CHECK-SAME to wrap the line

**Review by pifon2a — COMMENTED** [review]
Looks great! Thank you so much! Could you also add xla_hlo.GatherOp -> xla_lhlo.GatherOp conversion to `hlo_legalize_to_lhlo.cc`? it should be just one line to populate the patterns, one line in `map_hlo_to_lhlo_op.h` to map the ops and a test.

File: tensorflow/compiler/xla/service/mlir_gpu/tests/gather.hlo (hunk 1 above)
Comment (xinan-jiang) [reply tensorflow/compiler/xla/service/mlir_gpu/tests/gather.hlo]: done

**Review by pifon2a — APPROVED** [review]
git clang-format?

**Comment by gbaned** [comment]
@xinan-jiang Can you please check @joker-eph's comments and keep us posted. Thanks!

**Comment by timshen91** [comment]
Unassigning myself since @pifon2a is more familiar with this code.

**Pull request merged**